language/go/fileinfo.go

```diff
@@ -665,22 +543,34 @@ func matchesOS(os, value string) bool {
 // if they are negated.
 //
 // The remaining arguments describe the file being tested. All of these may
-// be empty or nil. osSuffix and archSuffix are filename suffixes. fileTags
-// is a list tags from +build comments found near the top of the file. cgoTags
+// be empty or nil. osSuffix and archSuffix are filename suffixes. tags
+// is the parsed build tags found near the top of the file. cgoTags
 // is an extra set of tags in a #cgo directive.
-func checkConstraints(c *config.Config, os, arch, osSuffix, archSuffix string, fileTags []tagLine, cgoTags tagLine) bool {
+func checkConstraints(c *config.Config, os, arch, osSuffix, archSuffix string, tags *buildTags, cgoTags *cgoTagsAndOpts) bool {
 	if osSuffix != "" && !matchesOS(os, osSuffix) || archSuffix != "" && archSuffix != arch {
 		return false
 	}
-	for _, l := range fileTags {
-		if !l.check(c, os, arch) {
-			return false
+
+	goConf := getGoConfig(c)
+	checker := func(tag string) bool {
+		if _, ok := rule.KnownOSSet[tag]; ok {
+			if os == "" {
+				return false
+			}
+			return matchesOS(os, tag)
+
+		} else if _, ok := rule.KnownArchSet[tag]; ok {
+			if arch == "" {
+				return false
+			}
+			return arch == tag
+
+		} else {
+			return goConf.genericTags[tag]
 		}
```

Comment:
nit: since we are returning early in each if block, we can probably rewrite this without using else

```
if foo {
return
}

if bar {
return
}

return
```